internal/orchestrator/reorg_handler_test.go

```diff
@@ -0,0 +1,502 @@
+		MainStorage:         mockMainStorage,
+		OrchestratorStorage: mockOrchestratorStorage,
+	}
+
+	mockRPC.EXPECT().GetChainID().Return(big.NewInt(1))
+	mockRPC.EXPECT().GetBlocksPerRequest().Return(rpc.BlocksPerRequestConfig{Blocks: 100})
+	mockOrchestratorStorage.EXPECT().GetLastReorgCheckedBlockNumber(big.NewInt(1)).Return(big.NewInt(100), nil)
+
+	mockMainStorage.EXPECT().LookbackBlockHeaders(big.NewInt(1), 10, big.NewInt(109)).Return([]common.BlockHeader{
+		{Number: big.NewInt(109), Hash: "hash109", ParentHash: "hash108"},
+		{Number: big.NewInt(108), Hash: "hash108", ParentHash: "hash107"}, // <-- fork ends here
+		{Number: big.NewInt(107), Hash: "hash107a", ParentHash: "hash106a"},
+		{Number: big.NewInt(106), Hash: "hash106a", ParentHash: "hash105a"},
+		{Number: big.NewInt(105), Hash: "hash105a", ParentHash: "hash104a"},
+		{Number: big.NewInt(104), Hash: "hash104a", ParentHash: "hash103a"},
+		{Number: big.NewInt(103), Hash: "hash103a", ParentHash: "hash102a"}, // <-- fork starts here
+		{Number: big.NewInt(102), Hash: "hash102", ParentHash: "hash101"},
+		{Number: big.NewInt(101), Hash: "hash101", ParentHash: "hash100"},
+		{Number: big.NewInt(100), Hash: "hash100", ParentHash: "hash99"},
+	}, nil)
+
+	mockRPC.EXPECT().GetBlocks([]*big.Int{big.NewInt(108), big.NewInt(107), big.NewInt(106), big.NewInt(105), big.NewInt(104), big.NewInt(103), big.NewInt(102), big.NewInt(101), big.NewInt(100)}).Return([]rpc.GetBlocksResult{
+		{BlockNumber: big.NewInt(108), Data: common.Block{Hash: "hash108", ParentHash: "hash107"}},
+		{BlockNumber: big.NewInt(107), Data: common.Block{Hash: "hash107", ParentHash: "hash106"}},
+		{BlockNumber: big.NewInt(106), Data: common.Block{Hash: "hash106", ParentHash: "hash105"}},
+		{BlockNumber: big.NewInt(105), Data: common.Block{Hash: "hash105", ParentHash: "hash104"}},
+		{BlockNumber: big.NewInt(104), Data: common.Block{Hash: "hash104", ParentHash: "hash103"}},
+		{BlockNumber: big.NewInt(103), Data: common.Block{Hash: "hash103", ParentHash: "hash102"}},
+		{BlockNumber: big.NewInt(102), Data: common.Block{Hash: "hash102", ParentHash: "hash101"}},
+		{BlockNumber: big.NewInt(101), Data: common.Block{Hash: "hash101", ParentHash: "hash100"}},
+		{BlockNumber: big.NewInt(100), Data: common.Block{Hash: "hash100", ParentHash: "hash99"}},
+	})
+
+	mockRPC.EXPECT().GetFullBlocks([]*big.Int{big.NewInt(103), big.NewInt(104), big.NewInt(105), big.NewInt(106), big.NewInt(107), big.NewInt(108)}).Return([]rpc.GetFullBlockResult{
+		{BlockNumber: big.NewInt(103), Data: common.BlockData{}},
+		{BlockNumber: big.NewInt(104), Data: common.BlockData{}},
+		{BlockNumber: big.NewInt(105), Data: common.BlockData{}},
+		{BlockNumber: big.NewInt(106), Data: common.BlockData{}},
+		{BlockNumber: big.NewInt(107), Data: common.BlockData{}},
+		{BlockNumber: big.NewInt(108), Data: common.BlockData{}},
+	})
+
+	mockMainStorage.EXPECT().DeleteBlockData(big.NewInt(1), mock.MatchedBy(func(blocks []*big.Int) bool {
+		return len(blocks) == 6
```

Comment:
why do we choose to re-insert the "fork end" block as well?
in this case block `(108), Hash: "hash108", ParentHash: "hash107"` can actually stay as it, no?

Reply:
(applies to the actual logic, if that's the case everywhere where we assert the `blocks []` argument `length` will be `1` element less)

Reply:
If you think about the reorg check as a moving window (which can be applied to the saved blocks at any time), then the latest block is not always the chain head.
So there can be a case, where the "fork end" is not actually the fork end and contains invalid data. So we just in case also check that block again. In the grand scheme it shouldn't make a big difference, it's just 1 block more to check whenever we detect a reorg

Reply:
got it, alright

the safer the better